Makes sure that substring outputs the entire string (or what's left of it), instead of "", if the specified length is to great.

andersjonsson opened the pull request:
if specified length is longer than input length, set length to the maximum possible length

fixes #141

```	
var people = new List<object>
{new {Name = "Long John", City = "New York"}, new {Name = "Short Mary", City = "Massachusetts"},};

Smart.Format("{Name:substr(5)}", people.First());
// result: "John"

Smart.Format("{City:substr(0,3)}", people.First());
// result: "New"

Smart.Format("{City:substr(999,0)}", people.First());
// result: ""

Smart.Format("{City:substr(0,999)}", people.First());
// result: "New York"
```

axunonb commented:
Thanks for the PR.
This would mean a breaking change, also regarding the [corresponding unit test](https://github.com/axuno/SmartFormat/blob/e803f08ad7a630eb621aec73a03859ea71c166b3/src/SmartFormat.Tests/Extensions/SubStringFormatterTests.cs#L40).
I see 3 options:
* Return an empty string (as it is implemented)
* Return a string from start to end of string, if length is longer (as in the PR)
* Throw an `ArgumentOutOfRangeException` following the C# string.SubString behavior. Maybe the "cleanest" option

What do you think?

andersjonsson commented:
I think SmartFormat should avoid exceptions when it comes to formatting. I agree that it is a breaking change, but it was also a behavior that was not covered by tests previously.

The two previous tests for length only tested cases where startpos was greater than input length.

andersjonsson commented:
The current implementation reasons that if you specify a starting position within the length of the input string,
and specify a length that is too great, you probably want the remainder of the string from that starting position
But, if you specify a starting position that is too great you get ""

example:
"New York"
//(0,10) => "New York"
//(4,10) => "York"
//(10,10) => ""

axunonb commented:
Good points indeed.
The extension already has some configuration inside.
We could add an `enum` for `OutOfRangeBevior` like `EmptyString`, `MaxStringLength`, `ThrowException` and avoid a breaking change (`MaxStringLength` should have a better wording, shouldn't it)

andersjonsson commented:
Good idea!

Yeah, naming it always hard 😉 
I suspect that ```StringRemainder``` would be misinterpreted and ```RestOfString``` sounds iffy
I don't know, you decide 😆

axunonb commented:
```
public enum OutOfRangeBehavior
{
    ReturnEmptyString,
    ReturnStartToEndOfString,
    ThrowException
}
```
:wink:

andersjonsson commented:
So
```
--- ReturnEmptyString ---
"New York"
//(0,10) => ""
//(4,10) => ""
//(10,10) => ""

--- ReturnStartToEndOfString ---
"New York"
//(0,10) => "New York"
//(4,10) => "York"
//(10,10) => ""

--- ThrowException ---
//(0,10) => OutOfRangeException
//(4,10) => OutOfRangeException
//(10,10) => OutOfRangeException
```
I'm not sure about the use cases for 1 and 3. Would the naming of 2 suggest that all of them should return "New York"?
I mean, "Start" could be interpreted as start of string *or* the submitted starting index.

I know I'm biased since I submitted this PR, but I think behavior 2 is a reasonable default and I'm not sure that 1 and 3 are useful.
What do you think? I suppose you have a much better understanding of how the lib is used "in the wild"


And how would I go about implementing it?
Just add a property to the SubStringFormatter?
If so, what's the best way of setting it runtime? Using GetFormatterExtension<SubStringFormatter>() and set the property that way?

axunonb commented:
> I mean, "Start" could be interpreted as start of string or the submitted starting index.
Then it should rather read `ReturnStartIndexToEndOfString` to be less ambiguous

> I think behavior 2 is a reasonable default and I'm not sure that 1 and 3 are useful.
Targeting to "no breaking change" would mean `ReturnEmptyString`  was the default.

> Just add a property to the SubStringFormatter?
Would be the most simple way, and initialize to default behavior.

>  Using GetFormatterExtension() and set the property that way?
Yep

axunonb commented:
Very cool, thanks a lot! Will create a new version next 2 hrs

andersjonsson commented:
aww chucks. I pushed a change to the tests a little too late ;)


Not that important, but add it if you want
I just split the test, and added a test for the ThrowException one

```
[Test]
public void LengthLongerThanString_ReturnEmptyString()
{
    var formatter = _smart.GetFormatterExtension<SubStringFormatter>()!;
    var behavior = formatter.OutOfRangeBehavior;
            
    formatter.OutOfRangeBehavior = SubStringFormatter.SubStringOutOfRangeBehavior.ReturnEmptyString;
    Assert.AreEqual(string.Empty, _smart.Format("{Name:substr(0,999)}", _people.First()));

    formatter.OutOfRangeBehavior = behavior;
}

[Test]
public void LengthLongerThanString_ReturnStartIndexToEndOfString()
{
    var formatter = _smart.GetFormatterExtension<SubStringFormatter>()!;
    var behavior = formatter.OutOfRangeBehavior;

    formatter.OutOfRangeBehavior = SubStringFormatter.SubStringOutOfRangeBehavior.ReturnStartIndexToEndOfString;
    Assert.AreEqual("Long John", _smart.Format("{Name:substr(0,999)}", _people.First()));

    formatter.OutOfRangeBehavior = behavior;
}

[Test]
public void LengthLongerThanString_ThrowException()
{
    var formatter = _smart.GetFormatterExtension<SubStringFormatter>()!;
    var behavior = formatter.OutOfRangeBehavior;

    formatter.OutOfRangeBehavior = SubStringFormatter.SubStringOutOfRangeBehavior.ThrowException;
    Assert.Throws<FormattingException>(() => _smart.Format("{Name:substr(0,999)}", _people.First()));

    formatter.OutOfRangeBehavior = behavior;
}
```

axunonb commented:
Added description to the [wiki](https://github.com/axuno/SmartFormat/wiki/SubString)
